### PR TITLE
Feature: Add xlm, eos, luna, xem as memo based currencies

### PR DESCRIPTION
**Pull request by xenirio**

### CODE REVIEW

File: web/src/components/Beneficiaries/BeneficiariesAddModal.tsx [hunk 1]
```diff
@@ -74,7 +75,8 @@ const BeneficiariesAddModalComponent: React.FC<Props> = (props: Props) => {
     const createSuccess = useSelector(selectBeneficiariesCreateSuccess);
     const createError = useSelector(selectBeneficiariesCreateError);
     const currencyItem = React.useMemo(() => currencies.find(item => item.id === currency), [currencies]);
-    const isRipple = React.useMemo(() => currency === 'xrp', [currency]);
+    const isDestinationTagExists = React.useMemo(() => currency === 'xrp', [currency]);
+    const isMemoTagExists = React.useMemo(() => CURRENCIES_MEMO.includes(currency), [currency]);
```
Comment (josadcha) [inline web/src/components/Beneficiaries/BeneficiariesAddModal.tsx:79]: move this to helpers, they're going to be reusable

Comment (josadcha) [reply web/src/components/Beneficiaries/BeneficiariesAddModal.tsx:79]: requiresTag
requiresMemo

File: web/src/components/Beneficiaries/BeneficiariesAddModal.tsx [hunk 2]
```diff
@@ -115,13 +117,27 @@ const BeneficiariesAddModalComponent: React.FC<Props> = (props: Props) => {
...
+            if (isDestinationTagExists) {
+                return `${coinAddress}?dt=${coinDestinationTag}`;
+            }
+
+            if (isMemoTagExists) {
+                return `${coinAddress}?memo=${coinDestinationTag}`;
+            }
+        } else {
+            return coinAddress;
+        }
```
Comment (josadcha) [inline web/src/components/Beneficiaries/BeneficiariesAddModal.tsx:131]:
```suggestion
        const address = requiresTag(currency) && coinDestinationTag ? `${rid}?dt=${tagMemo}`
            : requiresMemo(currency) && coinDestinationTag ? `${rid}?memo=${tagMemo}`
            : rid;
```

no need to make such a big function

File: web/src/components/Beneficiaries/BeneficiariesAddModal.tsx [hunk 3]
```diff
@@ -222,8 +238,24 @@ const BeneficiariesAddModalComponent: React.FC<Props> = (props: Props) => {
     const handleChangeFieldValue = React.useCallback((key: string, value: string) => {
         switch (key) {
             case 'coinAddress':
-                setCoinAddress(value);
-                validateCoinAddressFormat(value);
+                if ((isDestinationTagExists || isMemoTagExists) && value.includes('?')) {
```
Comment (josadcha) [inline web/src/components/Beneficiaries/BeneficiariesAddModal.tsx:241]: add please these functions to helpers

```
export const getMemoTag = (value?: string) => {
    if (value && (value.includes('?dt=') || value.includes('?memo='))) {
        return value.substr(value.lastIndexOf('=') + 1);
    }

    return '-';
};

export const getAddressWithoutMemoTag = (value?: string) => {
    if (value && (value.includes('?dt=') || value.includes('?memo='))) {
        return value.slice(0, value.lastIndexOf('?'));
    }

    return value;
};

```

**Review by josadcha — CHANGES_REQUESTED** [review]

File: web/src/helpers/memoBasedAsset.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+export const requiresTag = (currency: string) => {
+    return React.useMemo(() => currency === 'xrp', [currency]);
+};
+
+export const requiresMemo = (currency: string) => {
+    return React.useMemo(() => CURRENCIES_MEMO.includes(currency), [currency]);
+};
+
+export const getMemoTag = (value?: string) => {
```
Comment (calj) [inline web/src/helpers/memoBasedAsset.ts:12]: dt means "destination tag", please rename

File: web/src/constants/index.ts [hunk 5]
```diff
@@ -136,3 +136,4 @@ export const colors = {
 
 export const FIXED_VOL_PRECISION = 2;
 export const DEFAULT_PERCENTAGE_PRECISION = 2;
+export const CURRENCIES_MEMO = ['xlm', 'eos', 'luna', 'xem'];
```
Comment (calj) [inline web/src/constants/index.ts:139]: please define a constant CURRENCIES_DT, even if there is only xrp in it

**Review by calj — APPROVED** [review]

File: web/src/components/Beneficiaries/BeneficiariesAddModal.tsx [hunk 6]
```diff
@@ -74,7 +75,8 @@ const BeneficiariesAddModalComponent: React.FC<Props> = (props: Props) => {
     const createSuccess = useSelector(selectBeneficiariesCreateSuccess);
     const createError = useSelector(selectBeneficiariesCreateError);
     const currencyItem = React.useMemo(() => currencies.find(item => item.id === currency), [currencies]);
-    const isRipple = React.useMemo(() => currency === 'xrp', [currency]);
+    const isDestinationTagExists = requiresDTTag(currency) === true;
+    const isMemoTagExists = requiresMemoTag(currency) === true;
```
Comment (akushniruk) [inline web/src/components/Beneficiaries/BeneficiariesAddModal.tsx:79]: change: ```requiresDTTag(currency) === true``` to ```requiresDTTag(currency)```

**Review by akushniruk — APPROVED** [review]

**Review by josadcha — APPROVED** [review]

**Pull request merged**